drivers/serial/uart_async_rx.c

```diff
@@ -0,0 +1,134 @@
+/*
+ * Copyright (c) 2023 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <string.h>
+#include <zephyr/drivers/uart.h>
+#include <zephyr/drivers/serial/uart_async_rx.h>
+
+static uint8_t inc(struct uart_async_rx *rx_data, uint8_t val)
+{
+	return (val + 1) & (rx_data->config->buf_cnt - 1);
+}
+
+static struct uart_async_rx_buf *get_buf(struct uart_async_rx *rx_data, uint8_t idx)
+{
+	uint8_t *p = rx_data->config->buffer;
+
+	p += idx * (rx_data->buf_len + sizeof(struct uart_async_rx_buf));
+
+	return (struct uart_async_rx_buf *)p;
+}
+
+uint8_t *uart_async_rx_buf_req(struct uart_async_rx *rx_data)
+{
+	uint8_t *data = NULL;
+
+	if (rx_data->free_buf_cnt != 0) {
+		struct uart_async_rx_buf *buf = get_buf(rx_data, rx_data->drv_buf_idx);
+
+		data = buf->buffer;
+		rx_data->drv_buf_idx = inc(rx_data, rx_data->drv_buf_idx);
+
+		atomic_dec(&rx_data->free_buf_cnt);
+	}
+
+	return data;
+}
+
+void uart_async_rx_on_rdy(struct uart_async_rx *rx_data, uint8_t *buffer, size_t length)
+{
+	/* Cannot use CONTAINER_OF because validation fails due to type mismatch:
```

Comment:
I have a small suggestion. For API functions, it's worth adding `__ASSERT_NO_MSG` to check pointers.